pkg/controllers/nsxserviceaccount/nsxserviceaccount_controller.go

```diff
@@ -0,0 +1,241 @@
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (r *NSXServiceAccountReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	obj := &nsxvmwarecomv1alpha1.NSXServiceAccount{}
+	log.Info("reconciling CR", "nsxserviceaccount", req.NamespacedName)
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerSyncTotal, MetricResType)
+
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		log.Error(err, "unable to fetch NSXServiceAccount CR", "req", req.NamespacedName)
+		return ResultNormal, client.IgnoreNotFound(err)
+	}
+
+	// Since NSXServiceAccount service can only be activated from NSX 4.1.0 onwards,
+	// So need to check NSX version before starting NSXServiceAccount reconcile
+	if !r.Service.NSXClient.NSXCheckVersionForNSXServiceAccount() {
+		err := errors.New("NSX version check failed, NSXServiceAccount feature is not supported")
+		updateFail(r, &ctx, obj, &err)
+		// if NSX version check fails, it will be put back to reconcile queue and be reconciled after 5 minutes
+		return ResultRequeueAfter5mins, nil
+	}
+
+	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateTotal, MetricResType)
+		if !controllerutil.ContainsFinalizer(obj, servicecommon.NSXServiceAccountFinalizerName) {
+			controllerutil.AddFinalizer(obj, servicecommon.NSXServiceAccountFinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "add finalizer", "nsxserviceaccount", req.NamespacedName)
+				updateFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			log.V(1).Info("added finalizer on CR", "nsxserviceaccount", req.NamespacedName)
+		}
+
+		if obj.Status.Phase == nsxvmwarecomv1alpha1.NSXServiceAccountPhaseRealized {
+			return ResultNormal, nil
+		}
+		if err := r.Service.CreateOrUpdateNSXServiceAccount(ctx, obj); err != nil {
+			if errors.As(err, &util2.RestrictionError{}) {
+				log.Error(err, err.Error(), "nsxserviceaccount", req.NamespacedName)
+				updateFail(r, &ctx, obj, &err)
+				return ResultNormal, nil
+			}
+			log.Error(err, "operate failed, would retry exponentially", "nsxserviceaccount", req.NamespacedName)
+			updateFail(r, &ctx, obj, &err)
+			return ResultRequeue, err
+		}
+		updateSuccess(r, &ctx, obj)
+	} else {
+		if controllerutil.ContainsFinalizer(obj, servicecommon.NSXServiceAccountFinalizerName) {
+			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, MetricResType)
+			if err := r.Service.DeleteNSXServiceAccount(ctx, types.NamespacedName{
+				Namespace: obj.Namespace,
+				Name:      obj.Name,
+			}); err != nil {
+				log.Error(err, "removing finalizer failed, would retry exponentially", "nsxserviceaccount", req.NamespacedName)
```

Comment:
"removing finalizer failed, would retry exponentially" -> "deleting failed, would retry exponentially" is OK, since nsxserviceaccount is mentioned after this string.